### PR TITLE
Improve ipv6 support

### PR DESCRIPTION
**Pull request by underrun**
**What changed?**

* specify pprof host to override localhost default
* use net.JoinHostPort instead of fmt.Sprintf
* lookup `localhost` in various code paths
* add a test for ringpop using ipv6
* fixes comments to remove claim of ipv4 only

**Why?**

Because temporal works with ipv6, but there are some clarifications in documentation and additional settings that could make support cleaner.

**How did you test it?**

locally

**Potential risks**

none known

**Is hotfix candidate?**

no

### CODE REVIEW

File: common/rpc/rpc.go [hunk 1]
```diff
@@ -162,7 +162,21 @@ func getListenIP(cfg *config.RPC, logger log.Logger) net.IP {
...
+			return net.IPv4(127, 0, 0, 1)
+		}
+		var listenIp net.IP
+		for _, ip := range ips {
+			listenIp = ip
+			if listenIp.To4() != nil {
+				return listenIp
+			}
+		}
+		return listenIp
```
Comment (MichaelSnowden) [inline common/rpc/rpc.go:179]: I think we should extract a common function for this, and I think it should have its own unit test

File: common/membership/ringpop/factory_test.go [hunk 2]
```diff
@@ -199,6 +203,10 @@ func (s *RingpopSuite) TestRingpopInvalidTLS() {
 	s.Error(runRingpopTLSTest(&s.Suite, s.insecureFactory, s.serverTLSFactoryB))
 }
 
+func (s *RingpopSuite) TestRingpopIPv6TLS() {
+	s.Error(runRingpopTLSTest(&s.Suite, s.serverIPv6TLSFactoryA, s.serverIPv6TLSFactoryB))
```
Comment (MichaelSnowden) [inline common/membership/ringpop/factory_test.go:207]: Why do we expect this to cause an error? What about a test for if the TLS config is valid for IPV6?

File: common/membership/ringpop/factory_test.go [hunk 3]
```diff
@@ -60,17 +61,20 @@ type (
...
+		serverIPv6TLSFactoryA *factory
+		serverIPv6TLSFactoryB *factory
 
 		insecureFactory *factory
 	}
 )
 
 const (
 	localhostIPv4 = "127.0.0.1"
+	localhostIPv6 = "0::1"
```
Comment (alfred-landrum) [inline common/membership/ringpop/factory_test.go:77]: What's the extra zero indicate? I've always seen this as:
```suggestion
	localhostIPv6 = "::1"
```

File: common/membership/ringpop/factory_test.go [hunk 4]
```diff
@@ -208,12 +216,16 @@ func runRingpopTLSTest(s *suite.Suite, serverA *factory, serverB *factory) error
 
 	// Ping A through B to make sure B's dialer uses TLS to communicate with A
 	hostPortA := chA.PeerInfo().HostPort
+	fmt.Printf("hostPortA: %s", hostPortA)
```
Comment (alfred-landrum) [inline common/membership/ringpop/factory_test.go:219]:
```suggestion
```

File: common/membership/ringpop/factory_test.go [hunk 5]
```diff
@@ -208,12 +216,16 @@ func runRingpopTLSTest(s *suite.Suite, serverA *factory, serverB *factory) error
...
 	if err := chB.Ping(context.Background(), hostPortA); err != nil {
 		return err
 	}
 
 	// Confirm that A's listener is actually using TLS
 	clientTLSConfig, err := serverB.TLSFactory.GetInternodeClientConfig()
+
+	fmt.Printf("clientTLSConfig: %v", clientTLSConfig)
+
```
Comment (alfred-landrum) [inline common/membership/ringpop/factory_test.go:228]:
```suggestion
```

File: common/membership/ringpop/factory_test.go (hunk 3 above)
Comment (underrun) [reply common/membership/ringpop/factory_test.go:77]: there are lots of ways to write it - the zeros (127 bits worth) are implied. but i've seen some IP parsers accept `0::1` and not `::1` for some reason. we can make this `::1` tho - should be fine either way.

File: common/rpc/rpc.go (hunk 1 above)
Comment (underrun) [reply common/rpc/rpc.go:179]: this sounds good - best place for this seems like `environment/env.go` right? we currently have a `Localhost` constant in there set to `127.0.0.1` which is mostly used in tests but would be nice to also have updated.

Comment (underrun) [reply common/rpc/rpc.go]: ok - this is moved to `environment/env.go`

File: common/membership/ringpop/factory_test.go (hunk 2 above)
Comment (underrun) [reply common/membership/ringpop/factory_test.go:206]: Its not so much that we expect it to error as much as that it confirms that tchannel supports ipv6. the question "are we sure ringpop supports ipv6" was one that was a little bit difficult to answer without just starting up a server using ipv6.

The test itself exercises `ringpop.factory.getListenIP` with `BindOnIP` set to an ipv6 address via `getTChannel()`, confirms that tchannel `Ping` works with ipv6, and also confirms that `tls.Dial` can make a tls connection over ipv6.

Basically it's an easy way to add one test and get some assurance about a lot of codepaths and their capability of handling an ipv6 address.

If there's a better way to do that I'm open.

File: environment/env.go [hunk 6]
```diff
@@ -26,13 +26,16 @@ package environment
...
 import (
 	"fmt"
+	"net"
 	"os"
 	"strconv"
 )
 
 const (
-	// Localhost default localhost
-	Localhost = "127.0.0.1"
```
Comment (underrun) [inline environment/env.go:35]: effectively `Localhost` is removed in favor of requiring a call to `GetLocalhostIP()` and we're also adding an env var for LOCALHOST_IP which feels a little odd but being explicit about which localhost ip address is required in dual stack ipv4/ipv6 envrionments because golang wil always choose ipv4 by default.

File: environment/env.go [hunk 7]
```diff
@@ -133,11 +143,41 @@ func SetupEnv() {
 	}
 }
 
+func lookupLocalhostIP(domain string) string {
```
Comment (underrun) [inline environment/env.go:146]: this is really only parameterized for testing. is there a better way to do that?

File: common/membership/ringpop/factory_test.go (hunk 2 above)
Comment (underrun) [reply common/membership/ringpop/factory_test.go:206]: interestingly this is failing in the build pipeline - i guess because the runner doesn't support ipv6?

Comment (underrun) [reply common/membership/ringpop/factory_test.go:206]: so two questions -
* do we want to be able to run tests that require ipv6
* do we want to narrow the scope of the test (if so, to what?)

Comment (underrun) [reply common/membership/ringpop/factory_test.go:206]: oh man - yeah i read this wrong. the intention is `NoError` here

File: environment/env_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLookupLocalhostIPSuccess(t *testing.T) {
+	a := assert.New(t)
```
Comment (MichaelSnowden) [inline environment/env_test.go:33]: nit: add t.Parallel()

**Review by MichaelSnowden — APPROVED** [review]

File: environment/env_test.go (hunk 8 above)
Comment (underrun) [reply environment/env_test.go:34]: added
